app/src/main/java/com/github/h3lp3rs/h3lp/database/FireDatabase.kt

```diff
@@ -0,0 +1,161 @@
+        db.child(key).setValue(value)
+    }
+
+    /**
+     * Gets a double from the database
+     * @param key The key in the database
+     * @return Future of double
+     */
+    override fun getDouble(key: String): CompletableFuture<Double> {
+        return get(key)
+    }
+
+    /**
+     * Sets a double to the database
+     * @param key The key in the database
+     * @param value The value of the double
+     */
+    override fun setDouble(key: String, value: Double) {
+        db.child(key).setValue(value)
+    }
+
+    /**
+     * Gets an int from the database
+     * @param key The key in the database
+     * @return Future of int
+     */
+    override fun getInt(key: String): CompletableFuture<Int> {
+        // NOTE: We have to recode this case as Firebase natively supports
+        // Longs and not Ints.
+        return get<Long>(key).thenApply { it.toInt() }
+    }
+
+    /**
+     * Sets an int to the database
+     * @param key The key in the database
+     * @param value The value of the int
+     */
+    override fun setInt(key: String, value: Int) {
+        db.child(key).setValue(value)
+    }
+
+    /**
+     * Applies an arbitrary action when the value associated to the value changes
+     * @param key The key in the database
+     * @param type The type of the value associated to the key
+     * @param action The action taken at change
+     */
+    override fun <T> addListener(key: String, type: Class<T>, action: (T) -> Unit) {
+        val l = object : ValueEventListener {
+            override fun onDataChange(snapshot: DataSnapshot) {
+                val v: T = if(type == String::class.java || type == Int::class.java ||
+                    type == Double::class.java || type == Boolean::class.java) {
+                    snapshot.getValue(type)!!
+                } else {
+                    val gson = Gson()
+                    gson.fromJson(snapshot.getValue(String::class.java)!!, type)
+                }
+                action(v)
+            }
+
+            override fun onCancelled(databaseError: DatabaseError) {
+                println("Firebase listener error: ${databaseError.toException()}")
+            }
+        }
+        db.child(key).addValueEventListener(l)
+        // Enrich the list & add to map
+        val ls = openListeners.getOrDefault(key, emptyList()) + listOf(l)
+        openListeners[key] = ls
+    }
+
+    /**
+     * Clears all listeners related to a given key
+     * @param key The key in the database
+     */
+    override fun clearListeners(key: String) {
+        val ls = openListeners.getOrDefault(key, emptyList())
+        for(l in ls) {
+            db.child(key).removeEventListener(l)
+        }
+        openListeners.remove(key)
+    }
+
+    /**
+     * Deletes an entry of a given key from the database
+     * @param key They key in the database
+     */
+    override fun delete(key: String) {
+        db.child(key).removeValue()

```

Comment:
I am not at all sure about this but shouldn't deleting a key also delete its listeners in openListeners by removing the key (I guess that the respective listeners on the actual database are deleted if you delete a key so there could be an inconsistency)?

Reply:
Nice catch, we implemented deletion before the listeners, so we just forgot it